File: E2E/check_embedlinear_weights.py
"""
Compare embedLinear weight VALUES between Python file and what Swift loaded
"""
import torch
from safetensors.torch import load_file
from pathlib import Path

PROJECT_ROOT = Path(__file__).parent.parent
MODELS_DIR = PROJECT_ROOT / "models" / "chatterbox"

# Load Python weights from file
flow_weights = load_file(str(MODELS_DIR / "s3gen.safetensors"))
python_weight = flow_weights["flow.encoder.embed.out.0.weight"]  # [512, 512]

print("=" * 80)
print("PYTHON EMBEDLINEAR WEIGHT (from file)")
print("=" * 80)
print(f"Shape: {python_weight.shape}")
print(f"PyTorch Linear format: [out_features, in_features] = [512, 512]")
print(f"Sum: {python_weight.sum().item():.6f}")
print(f"Mean: {python_weight.mean().item():.6f}")
print(f"Std: {python_weight.std().item():.6f}")
print(f"Range: [{python_weight.min().item():.6f}, {python_weight.max().item():.6f}]")
print(f"\nFirst 5 values [0, :5]: {python_weight[0, :5].tolist()}")
print(f"First 5 values [:5, 0]: {python_weight[:5, 0].tolist()}")

print("\n" + "=" * 80)
print("WHAT SWIFT SHOULD HAVE AFTER TRANSPOSE")
print("=" * 80)
print("MLX Linear format: [in_features, out_features]")
print("So we TRANSPOSE: [512, 512] -> [512, 512]")
transposed = python_weight.T
print(f"Transposed sum: {transposed.sum().item():.6f} (should match Python: 6.151423)")
print(f"\nTransposed first 5 values [0, :5]: {transposed[0, :5].tolist()}")
print(f"Transposed first 5 values [:5, 0]: {transposed[:5, 0].tolist()}")

File: swift/test_scripts/CheckEmbedLinearWeights.swift
import Foundation
import MLX
import Nightingale

@main
struct CheckEmbedLinearWeights {
    static func main() async throws {
        print("Loading ChatterboxEngine...")
        let modelsPath = URL(fileURLWithPath: "../models/chatterbox")
        let engine = ChatterboxEngine()
        try await engine.loadModels(modelsURL: modelsPath)

        guard let s3gen = await engine.s3gen else {
            fatalError("S3Gen not loaded!")
        }

        let w = s3gen.encoder.embedLinear.weight
        eval(w)

        print("=" + String(repeating: "=", count: 79))
        print("SWIFT EMBEDLINEAR WEIGHT (after loading via ChatterboxEngine)")
        print("=" + String(repeating: "=", count: 79))
        print("Shape: \(w.shape)")
        print("Sum: \(w.sum().item(Float.self))")
        print("Mean: \(w.mean().item(Float.self))")
        print("Std: \(w.variance().sqrt().item(Float.self))")
        print("Range: [\(w.min().item(Float.self)), \(w.max().item(Float.self))]")

        // Check first few VALUES to compare with Python
        print("\nFirst 5 values [0, :5]:")
        let row0 = w[0, 0..<5]
        eval(row0)
        for i in 0..<5 {
            print("  [\(i)]: \(row0[i].item(Float.self))")
        }

        print("\nFirst 5 values [:5, 0]:")
        let col0 = w[0..<5, 0]
        eval(col0)
        for i in 0..<5 {
            print("  [\(i)]: \(col0[i].item(Float.self))")
        }

        print("\n" + String(repeating: "=", count: 80))
        print("EXPECTED (Python transposed to MLX format [in, out]):")
        print(String(repeating: "=", count: 80))
        print("First 5 values [0, :5]: [-0.022813, 0.000403, -0.127293, -0.002905, 0.003286]")
        print("First 5 values [:5, 0]: [-0.022813, 0.015566, -0.000571, -0.013604, 0.017855]")
    }
}
